File: projects/mediavault/frontend/src/lib/rbac.ts
import type { WorkspaceRole } from '@/types'

const ROLE_RANK: Record<WorkspaceRole, number> = {
  VIEWER: 0,
  MEMBER: 1,
  ADMIN: 2,
}

export function roleAtLeast(role: WorkspaceRole | null | undefined, minimum: WorkspaceRole): boolean {
  if (!role) return false
  return ROLE_RANK[role] >= ROLE_RANK[minimum]
}

export function canWrite(role: WorkspaceRole | null | undefined): boolean {
  return roleAtLeast(role, 'MEMBER')
}

export function canAdminister(role: WorkspaceRole | null | undefined): boolean {
  return roleAtLeast(role, 'ADMIN')
}

export function roleLabel(role: WorkspaceRole): string {
  switch (role) {
    case 'ADMIN':
      return 'Admin'
    case 'MEMBER':
      return 'Member'
    case 'VIEWER':
      return 'Viewer'
  }
}

File: projects/mediavault/backend/app/services/rbac.py
"""Role-based access control helpers.

Role hierarchy (each role inherits the permissions of the ones below it):

    ADMIN > MEMBER > VIEWER

- VIEWER: read-only access to workspace assets, folders, tags, search, shares.
- MEMBER: VIEWER + create/update/delete their own assets, folders, tags, shares.
- ADMIN:  MEMBER + manage workspace settings, manage memberships, and act on
          any asset/folder/tag regardless of ownership.
"""

from app.models.membership import WorkspaceRole

_ROLE_RANK = {
    WorkspaceRole.VIEWER: 0,
    WorkspaceRole.MEMBER: 1,
    WorkspaceRole.ADMIN: 2,
}


def role_at_least(role: WorkspaceRole, minimum: WorkspaceRole) -> bool:
    return _ROLE_RANK[role] >= _ROLE_RANK[minimum]


def can_write(role: WorkspaceRole) -> bool:
    return role_at_least(role, WorkspaceRole.MEMBER)


def can_administer(role: WorkspaceRole) -> bool:
    return role_at_least(role, WorkspaceRole.ADMIN)


def can_manage_resource(role: WorkspaceRole, is_owner: bool) -> bool:
    """MEMBERs can manage resources they own; ADMINs can manage anything."""
    if can_administer(role):
        return True
    return can_write(role) and is_owner
